tensorflow/python/layers/maxout.py

```diff
@@ -0,0 +1,111 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+# pylint: disable=unused-import,g-bad-import-order
+"""Contains the maxout layer
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import gen_array_ops
+
+from tensorflow.python.layers import base
+
+
+def maxout(inputs, num_units, axis=-1, name=None):
+    """Adds a maxout op from https://arxiv.org/abs/1302.4389
+
+      "Maxout Networks"
+
+      Ian J. Goodfellow, David Warde-Farley, Mehdi Mirza, Aaron Courville, Yoshua
+      Bengio
+
+    Usually the operation is performed in the filter/channel dimension. This can also be
+    used after fully-connected layers to reduce number of features.
+
+    Args:
+      inputs: Tensor input
+      num_units: Specifies how many features will remain after maxout in the `axis` dimension (usually channel).
+      This must be multiple of number of `axis`.
+      axis: The dimension where max pooling will be performed. Default is the
+        last dimension.
+      name: Optional scope for name_scope.
+    Returns:
+      A `Tensor` representing the results of the pooling operation.
+    Raises:
+      ValueError: if num_units is not multiple of number of features.
+    """
+
+    return MaxOut(num_units=num_units, axis=axis, name=name)(inputs)
+
+
+class MaxOut(base.Layer):
+    """Adds a maxout op from https://arxiv.org/abs/1302.4389
+
+      "Maxout Networks"
+
+      Ian J. Goodfellow, David Warde-Farley, Mehdi Mirza, Aaron Courville, Yoshua
+      Bengio
+
+    Usually the operation is performed in the filter/channel dimension. This can also be
+    used after fully-connected layers to reduce number of features.
+
+    Args:
+      inputs: Tensor input
+      num_units: Specifies how many features will remain after maxout in the `axis` dimension (usually channel).
+      This must be multiple of number of `axis`.
+      axis: The dimension where max pooling will be performed. Default is the
+        last dimension.
+      name: Optional scope for name_scope.
+    Returns:
+      A `Tensor` representing the results of the pooling operation.
+    Raises:
+      ValueError: if num_units is not multiple of number of features.
+    """
+
+    def __init__(self,
+                 num_units,
+                 axis=-1,
+                 name=None,
+                 **kwargs):
+        super(MaxOut, self).__init__(
+            name=name, trainable=False, **kwargs)
+        self.axis = axis
+        self.num_units = num_units
+
+    def call(self, inputs, training=False):
+        inputs = ops.convert_to_tensor(inputs)
+        shape = inputs.get_shape().as_list()
+        if self.axis is None:
```

Comment:
This will only be None if the user explicitly passes None, which is a little weird. How about making the default value in the constructor None so -1 only appears in a single place in code?

Reply:
I'm back.

Yes, you are right. I'll patch that.